samples/valueentity-counter/src/main/java/com/example/domain/CounterImpl.java

```diff
@@ -21,45 +26,64 @@ public CounterImpl(@EntityId String entityId) { // <2>
 
     // tag::increase[]
     @Override
-    protected Empty increase(CounterApi.IncreaseValue command, CommandContext<CounterDomain.CounterState> ctx) {
+    public Effect<Empty> increase(
+            Optional<CounterDomain.CounterState> currentState,
+            CounterApi.IncreaseValue command,
+            CommandContext<Empty, CounterDomain.CounterState> context) {
         if (command.getValue() < 0) { // <1>
-            throw ctx.fail("Increase requires a positive value. It was [" + command.getValue() + "].");
+            return context.effectBuilder().failure("Increase requires a positive value. It was [" + command.getValue() + "].");
         }
-        CounterDomain.CounterState state = ctx.getState() // <2>
+        CounterDomain.CounterState state = currentState // <2>
                 .orElseGet(() -> CounterDomain.CounterState.newBuilder().build()); // <3>
         CounterDomain.CounterState newState =  // <4>
                 state.toBuilder().setValue(state.getValue() + command.getValue()).build();
-        ctx.updateState(newState); // <5>
-        return Empty.getDefaultInstance();
+        return context.effectBuilder()
+                .updateState(newState) // <5>
+                .thenReply(Empty.getDefaultInstance()); // FIXME add convenience shortcut for reply Empty?
```

Comment:
Yep, I was having the same thought. 

We should consider is if it realistic to think that one day we will support other formats than protobuf. If not (most likely not), it's fine to bound ourselves to `Empty`. 

An alternative is to have our own type (eg: `Done`) to replace `Empty`. And then promote the idea that if you just want to return an ack, you can use `Done` and the builder has a convenience for it. 

Anyway, not a big deal if we are sure about having a long term relationship with proto's `Empty`

Reply:
We already support json. I agree that this is just a convenience that can be added if needed.

Reply:
Rather than tying this to protobuf `Empty` specifically, it could be nicer if it returned a default instance of whatever the gRPC service method's return type is. I think that's supported in the JS SDK. Not sure how easy with the new approach, but possible if it has a handle on the method descriptor?

Reply:
That would be interesting, but I don't know if I understand what you mean with a handle on the method descriptor. 

If a method returns `Foo`, we need to return an `Effect<Foo>`. If the builder is not already typed on `Foo` we will need something like: `replyWith(Foo.class)` and later behind the scenes build a `Foo.getDefaultInstance()` on behalf of the user. 

Between ``replyWith(Foo.class)` and `thenReply(Foo.getDefaultInstance())`, we save some characters but we increase API surface for something that is not so hard for the users to do.

Reply:
I think it would work with `return effects().replyWithEmpty()`. The `Foo` can be inferred from the return type of the method. We would create an `EmptyReply<Foo>`, which we will translate to the `Foo.getDefaultInstance()`.

Reply:
and `Foo.getDefaultInstance()` would be constructed via the proto method from the gRPC service method that Peter mentioned.

Reply:
Yes, I was thinking of just `replyWithEmpty()` or `replyWithDefault()` or similar and the actual default instance created using the gRPC method descriptor.